File: Lab1_3/1.cpp
 //Комплексные числа
#include <iostream>
#include <complex>
#include <math.h>
#include <iomanip>
   
using namespace std;

int main() {
int n, k;
double real, imag, z_real, z_imag, n_rt, dn_rt;
complex< double > z1( 1.0, 1.0 );
complex< double > z2( 3.0, -1.0 );
complex< double > z3( -1.0, 1.0 );

cout <<"z1=" << z1 << endl; // Комплексное число выводится в виде вектора
cout << "z1^sopr=" << conj( z1 ) << endl; // Комплексно-сопряженное
cout <<"z2=" << z2 << endl; // Комплексное число выводится в виде вектора
cout << "z2^sopr=" << conj( z2 ) << endl; // Комплексно-сопряженное
cout << "Re z1=" << z1.real() << endl; // Действительная часть комплексного числа
cout << "Im z1=" << z1.imag() << endl; // Мнимая часть комплексного числа
cout << "Re z2=" << z2.real() << endl; // Действительная часть комплексного числа
cout << "Im z2=" << z2.imag() << endl; // Мнимая часть комплексного числа
cout << "z1 + z2=" << z1 + z2 << endl; // Сложение комплексных чисел
cout << "z1 - z2=" << z1 - z2 << endl; // Разность комплексных чисел
cout << "z1 * z2=" << z1 * z2 << endl; // Умножение комплексных чисел
cout << "z1 / z2=" << z1 / z2 << endl; // Деление комплексных чисел


cout << "z3^3=" << pow( z3, 4 ) << endl; // Возведение в степень комплексного числа

double rho = abs( z3 ); // Модуль комплексного числа
double phi = arg( z3 ); // Аргумент комплексного числа
// Тригонометрическая форма комплексного числа
cout << "z3=" << rho << "*(cos(" << phi << ") + i*sin(" << phi << "))" << endl;
// Показательная форма комплексного числа
cout <<"z3=" << rho << "*exp(i*" << phi << ")" << endl;
// Получаем исходное комплексное число по модулю и аргументу
cout << "z3=" << polar( rho, phi ) << endl;


// Извлечение корня n степени из комплексного числа
cout << "Input complex number (example: -1 1) " ;
cin >> real >> imag; // Ввод действительной и мнимой части через пробел
complex zp(real, imag);
cout << fixed << setprecision(16) << zp << endl;
cout << "Input number n_root " ;
cin >> n_rt;
cout << "\n";
rho = abs( zp );
phi = arg( zp );
cout << "rho=" << rho << ", phi=" << phi << endl;
dn_rt = 1.0/n_rt;
for(k=0; k<n_rt;k++){
    z_real = pow( rho, float (1/n_rt)) * cos((phi + 2 * M_PI * k) / n_rt);
    z_imag = pow( rho, float (1/n_rt)) * sin((phi + 2 * M_PI * k) / n_rt);
    complex z_n(z_real, z_imag);
    cout << "k=" << k << ", z_k=" << z_n << "\n";}


return 0;
} 

File: Lab1_3/1.py
# Комплексные числа, операции
import math
import cmath

# Функция для перевода комплексного числа в тригонометрическую форму
def complex_to_trig(z):
    r, theta = cmath.polar(z)
    return r, theta

# Функция для возведения комплексного числа в степень
def complex_pow(z, n):
    r, theta = complex_to_trig(z)
    return r**n * cmath.exp(1j * n * theta)

# Функция для извлечения корня из комплексного числа
def complex_root(z, n):
    r, theta = complex_to_trig(z)
    roots = []
    for k in range(n):
        root = r**(1/n) * cmath.exp(1j * (theta + 2 * math.pi * k) / n)
        roots.append(root)
    return roots

# Арифметические операции
z1 = complex(input("Введите комплексное число z1, формат a+bj: "))
z2 = complex(input("Введите комплексное число z2, формат a+bj: "))
print(f'z1+z2={z1+z2}, z1-z2={z1-z2}, z1*z2={z1*z2}, z1/z2={z1/z2}')

z3 = complex(input("Введите комплексное число z3, формат a+bj: "))
n = int(input("Введите степень, в которую возвести z3: "))
n1 = int(input("Введите степень корня: "))

r, theta = complex_to_trig(z3)
print("Комплексное число в тригонометрической форме:", f'{r:.5f}*(cos({theta:.5f}) + i*sin({theta:.5f})')
print("Возведение в степень:", complex_pow(z3, n))
print("Извлечение корня:", complex_root(z3, n1))
